### PR TITLE
Module UI generation from config files

### PR DESCRIPTION
**Pull request by user2589**
does not include js/css update, only server side code

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,7 +6,7 @@ test:
 	python -m easy_phi.tests.run_tests
 
 run:
-	easy_phi/app.py
+	python easy_phi/app.py
```
Comment (user2589) [inline Makefile:9]: to allow run from filesystems mounted with noexec

File: easy_phi/app.py [hunk 2]
```diff
@@ -33,16 +34,6 @@
 
 define("default_format", default='json')
 
-settings = {
```
Comment (user2589) [inline easy_phi/app.py:36]: moved to line 261 where configuration from config file and command line already parsed

File: easy_phi/app.py [hunk 3]
```diff
@@ -210,19 +236,16 @@ def get(self):
...
-        parse_command_line(final=False)
-        parse_config_file(options.conf_path, final=False)
-        # command-line args override conf file options
-        parse_command_line()
-    else:
-        try:
-            parse_config_file(options.conf_path, final=False)
-        except IOError:  # configuration file doesn't exist, use defaults
-            pass
+class ContorlledCacheStaticFilesHandler(tornado.web.StaticFileHandler):
```
Comment (user2589) [inline easy_phi/app.py:239]: this handler disables caching when working in debug mode. It allows to test js/css changes without app restart

File: easy_phi/app.py [hunk 4]
```diff
@@ -235,6 +258,22 @@ def main(application):
...
+    parse_config_file(options.conf_path, final=False)
+    # command-line args override conf file options
+    parse_command_line()
+else:
+    try:
+        parse_config_file(options.conf_path, final=False)
+    except IOError:  # configuration file doesn't exist, use defaults
+        pass
+
+settings = {
```
Comment (user2589) [inline easy_phi/app.py:273]: moving it here allows to pass right value of debug and enables autoreload. I.e. this way app automatically restarts if any of imported files was changed.

File: easy_phi/tests/scpi2widgets_test.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+[RESET]
+scpi = *RST
+widget = {reset_widget}
+
+[System name]
+scpi = SYSTem:NAME?
+widget = {sysname_widget}\n""".format(reset_widget=self.reset_widget,
+                                      sysname_widget=self.sysname_widget))
+        widgets_conf.flush()
+        self.conf = widgets_conf
```
Comment (user2589) [inline easy_phi/tests/scpi2widgets_test.py:30]: this line costs about 2 hours of work, yet does not make any sense on first look. Without it, widget_conf removed by garbage collector after setUp finished

File: static/index.html [hunk 6]
```diff
@@ -48,8 +50,8 @@
 </footer>
 
 <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
-<script src="//code.jquery.com/ui/1.11.4/jquery-ui.js"></script>
-<script src="js/main.js"></script>
+<script src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.4/jquery-ui.min.js"></script>
+<script src="/static/js/main.js"></script>
```
Comment (user2589) [inline static/index.html:54]: always use absolute paths for static files. If you move page to a different place, relative links will point to a different location.

File: static/index.html [hunk 7]
```diff
@@ -3,8 +3,10 @@
 <head>
     <title>Easy Phi - test measurement equipment platform</title>
     <meta content="text/html;charset=utf-8" http-equiv="Content-Type">
+    <link rel="stylesheet" type="text/css" href="http://yui.yahooapis.com/3.18.1/build/cssreset/cssreset-min.css">
     <link rel="stylesheet" href="css/style.css" type="text/css" media="screen, projection">
-    <link rel="stylesheet" href="css/jquery-ui.min.css" type="text/css" media="screen, projection">
```
Comment (user2589) [inline static/index.html:7]: previous local hosted theme was mising some image files and caused a lot of 404 errors.
This solution is temporary, ultimate goal is to get rid of  jQueryUI completely

File: easy_phi/tests/handlers_test.py [hunk 8]
```diff
@@ -97,12 +101,77 @@ def test_list_scpi_commands(self):
         self.failIf(response.error)
         response_obj = json.loads(response.body)
 
-        self.assertTrue(isinstance(response_obj, list),
-                        "module_scpi_list expected to return list of commands")
+        self.assertIsInstance(
+            response_obj, list,
+            "module_scpi_list expected to return list of commands")
```
Comment (mansuleman) [inline easy_phi/tests/handlers_test.py:106]: Test is failing for me:

```
======================================================================
FAIL: test_list_scpi_commands (easy_phi.tests.handlers_test.ListSCPICommandsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/tornado/testing.py", line 120, in __call__
result = self.orig_method(*args, **kwargs)
  File "easy_phi/tests/handlers_test.py", line 109, in test_list_scpi_commands
"module_scpi_list returned empty list")
AssertionError: module_scpi_list returned empty list
----------------------------------------------------------------------
```

upd: Please be careful with markdown

Comment (user2589) [inline easy_phi/tests/handlers_test.py:106]: You don't have configuration patch file installed
for debugging purposes, just do:

```
sudo ln -s <full path to repository>/scripts/modules_conf_patches /etc/easy_phi/modules_conf_patches.conf
```

or, run with parameter:

```
python easy_phi/app.py --modules_conf_patches_path=scripts/modules_conf_patches.conf
```

Comment (mansuleman) [inline easy_phi/tests/handlers_test.py:106]: The first command gives me:
<code>sudo ln -s scripts/modules_conf_patches /etc/easy_phi/modules_conf_patches.conf</code>
<code>ln: failed to create symbolic link ‘/etc/easy_phi/modules_conf_patches.conf’: No such file or directory
</code><br>

The second command just runs the app with the params but not tests, so the test is still failing.

**Comment by mansuleman** [comment]
Ship it

File: easy_phi/tests/handlers_test.py (hunk 8 above)
Comment (user2589) [inline easy_phi/tests/handlers_test.py:106]: scripts/blablah is not a full path :)

Comment (mansuleman) [inline easy_phi/tests/handlers_test.py:106]: this was fixed for me yesterday. see ship it sign

**Comment by rezniknn** [comment]
Ship it

**Pull request merged**